### PR TITLE
ApplySnapshot function : TODO implmented for handling ErrorSnapOutOfDate and test case added.

### PR DESCRIPTION
**Pull request by sinsharat**
raft/storage.go: Implemented TODO for returning ErrorSnapOutOfDate. Also test case added for the same.

This was a TODO requirement which was not implemented

### CODE REVIEW

File: Documentation/demo.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ On each etcd node, specify the cluster members:
 TOKEN=token-01
 CLUSTER_STATE=new
 NAME_1=machine-1
-NAME_2=machine-2
+NAME_2=machine-2
```
Comment (heyitsanthony) [inline Documentation/demo.md]: revert this?

File: raft/storage.go [hunk 2]
```diff
@@ -169,6 +169,11 @@ func (ms *MemoryStorage) ApplySnapshot(snap pb.Snapshot) error {
 	defer ms.Unlock()
 
 	// TODO: return ErrSnapOutOfDate?
```
Comment (heyitsanthony) [inline raft/storage.go]: comment can be removed now?

File: raft/storage_test.go [hunk 3]
```diff
@@ -256,3 +256,33 @@ func TestStorageAppend(t *testing.T) {
...
+	tests := []struct {
+		wsnap pb.Snapshot
+	}{
+		{pb.Snapshot{Data: data, Metadata: pb.SnapshotMetadata{Index: 4, Term: 4, ConfState: *cs}}},
+		{pb.Snapshot{Data: data, Metadata: pb.SnapshotMetadata{Index: 3, Term: 3, ConfState: *cs}}},
+	}
+
+	s := NewMemoryStorage()
+
+	//Apply Snapshot successfull
```
Comment (heyitsanthony) [inline raft/storage_test.go]: successful

**Comment by heyitsanthony** [comment]
rebase on a more recent version of master to clear the semaphore errors and change the commit title to conform with the style `pkg: commit description` to clear the travis error?

File: raft/storage.go [hunk 4]
```diff
@@ -169,6 +169,11 @@ func (ms *MemoryStorage) ApplySnapshot(snap pb.Snapshot) error {
...
+	msIndex := ms.snapshot.Metadata.Index
+	snapIndex := snap.Metadata.Index
+	if msIndex > snapIndex {
```
Comment (xiang90) [inline raft/storage.go]: msIndex >= snapIndex

File: raft/storage_test.go [hunk 5]
```diff
@@ -256,3 +256,33 @@ func TestStorageAppend(t *testing.T) {
 		}
 	}
 }
+
+func TestStorageApplySnapshot(t *testing.T) {
+	cs := &pb.ConfState{Nodes: []uint64{1, 2, 3}}
+	data := []byte("data")
+
+	tests := []struct {
```
Comment (gyuho) [inline raft/storage_test.go]: Why don't we just define `[]pb.Snapshot`?

File: raft/storage.go (hunk 4 above)
Comment (sinsharat) [inline raft/storage.go]: I had not added msIndex >= snapIndex assuming that if its the same as recent snap shot why to throw error for it. Is it ok to add msIndex >= snapIndex for this?

File: Documentation/demo.md (hunk 1 above)
Comment (sinsharat) [inline Documentation/demo.md]: ok, i will do once i re-submit this.

File: raft/storage.go (hunk 2 above)
Comment (sinsharat) [inline raft/storage.go]: ok will remove the comment in the re-submit.

File: raft/storage_test.go (hunk 3 above)
Comment (sinsharat) [inline raft/storage_test.go]: i will correct the spelling on next re-submit.

File: raft/storage_test.go (hunk 5 above)
Comment (sinsharat) [inline raft/storage_test.go]: ok will modify the test case accordingly to use this when i re-submit.

**Comment by sinsharat** [comment]
another pull request applied with ID #6099 with fixes

**Pull request closed without merging**